Redo priority queue

peterebden opened the pull request:
Once https://github.com/thought-machine/please/pull/1102 is merged, this has some more changes on top; it removes the central priority queue and acts on channels directly. 
Quite a few things are simplified & we should avoid some centralised contention.

peterebden committed:
Break out parse tasks into their own channel

They don't gain from the priority queue since we create a goroutine for each anyway. This should reduce contention on it a bit.

peterebden committed:
Remove these priorities

peterebden committed:
Rip out the central priority queue and just put things straight onto channels.

peterebden committed:
Simplify AddPendingTest from the outside

peterebden committed:
Prevent panics

peterebden committed:
Fix closure

peterebden committed:
Ensure we don't try to re-close channels.

peterebden committed:
gofmt

Tatskaari commented on src/core/state.go:
```diff
@@ -270,135 +234,94 @@ func (state *BuildState) addActiveTargets(n int) {
-		state.addPending(label, Build)
-	}
+func (state *BuildState) addPendingBuild(target *BuildTarget) {
+	atomic.AddInt64(&state.progress.numPending, 1)
```
How did we get away with not having this here before?

peterebden replied:
It got called within `addPendingTask`, but I ended up deleting that because it's not generic any more. We'll still be calling it the same number of times.

Tatskaari committed:
Allow lists for gentest()! (#1760)

Tatskaari committed:
Added go_module() codelab (#1761)

Tatskaari committed:
V16 milestone (#1762)

* WIP

* v16 migration guide

* Small fixes

Tatskaari committed:
Fix missing arm64 release artifacts (#1763)

Tatskaari committed:
Tag Please v16.0.0 (#1765)

* Tag Please v16.0.0

* Add changelogs for new commits

* Updated changelog

Tatskaari committed:
Fix issue with runtime hashes for local tests (#1769)

* Fix issue with runtime hashes for local tests

* nothing to see here

peterebden committed:
Try breaking it out of being a separate type (#1773)

Tatskaari committed:
Record xattrs for the hash when retrieving from the http cache (#1771)

Commit:
Template tweaks: fix invalid html, add opengraph tags (#1776)

* Meta charset should be inside the header tag

nothing should be outside the html tag

* Adding meta property opengraph tags

To make it prettier when shared on social media

* Swap green triangle for teal

Tatskaari committed:
Tag v16.0.1 (#1775)

tiagovtristao committed:
Add output grouping option to plz run parallel (#1768)

Add `--output` flag to `plz run parallel` for allowing for different output types:
- `--output=group_immediate` groups tasks' outputs and flushes them to the screen as they complete
- `--output=quiet` deprecates the `--quiet` flag

peterebden committed:
Implement per-function profiling metrics for asp (#1774)

* Implement per-function profiling metrics for asp

* slightly simpler here

peterebden committed:
Graph parallelism improvements (#1102)

<stuff>

Co-authored-by: jpoole <jpoole@thoughtmachine.net>

Tatskaari committed:
Fix plz watch --run process cleanup (#1780)

* Fix plz watch --run process cleanup

* Fix up test
